python/ray/experimental/channel/common.py

```diff
@@ -333,23 +355,67 @@ def start(self):
         pass
 
     def _read_list(self, timeout: Optional[float] = None) -> List[Any]:
-        results = []
-        for c in self._input_channels:
-            start_time = time.monotonic()
-            results.append(c.read(timeout))
-            if timeout is not None:
-                timeout -= time.monotonic() - start_time
-                timeout = max(timeout, 0)
+        # Release the buffers of the channels that were not read in the last
+        # `read` call because a RayTaskError was found. If we don't do this,
+        # the read operation will read different versions of the object refs.
+        if self._unread_channels is not None:
+            self._release_channel_buffers(self._unread_channels, timeout)
+            # Do not update `remaining_timeout` to avoid causing an unexpected
+            # timeout in the following read operation.
+            self._unread_channels = None
+
+        results = [None for _ in range(len(self._input_channels))]
+        if timeout is None or timeout == -1:
+            timeout = float("inf")
+        timeout_point = time.monotonic() + timeout
+        remaining_timeout = timeout
+
+        from ray.dag import DAGContext
+
+        ctx = DAGContext.get_current()
+        iteration_timeout = ctx.read_iteration_timeout_s
+
+        # Iterate over the input channels with a shorter timeout for each iteration
+        # to detect RayTaskError early and fail fast.
+        read_channel_set = set()
+        while len(read_channel_set) < len(self._input_channels):
+            for i, c in enumerate(self._input_channels):
+                if c in read_channel_set:
+                    continue
+                try:
+                    result = c.read(min(remaining_timeout, iteration_timeout))
+                    results[i] = result
+                    read_channel_set.add(c)
+                    if isinstance(result, ray.exceptions.RayTaskError):
+                        # If we raise an exception immediately, it will be considered
+                        # as a system error which will cause the execution loop to
+                        # exit. Hence, return immediately and let `_process_return_vals`
+                        # handle the exception.
+                        #
+                        # Return a list of RayTaskError so that the caller will not
+                        # get an undefined partial result.
+                        self._unread_channels = [
+                            c for c in self._input_channels if c not in read_channel_set
+                        ]
+                        fast_fail_error = ray.exceptions.RayChannelFastFailError(
+                            i, result
+                        )
+                        return [
```

Comment:
open an issue to provide a way to return the results which has already been read successfully.

Reply:
https://github.com/ray-project/ray/issues/49769